File: src/main/java/brightspark/landmanager/message/MessageCreateArea.java
package brightspark.landmanager.message;

import brightspark.landmanager.LandManager;
import brightspark.landmanager.data.areas.AddAreaResult;
import brightspark.landmanager.data.areas.Area;
import brightspark.landmanager.data.areas.CapabilityAreas;
import brightspark.landmanager.data.logs.AreaLogType;
import brightspark.landmanager.event.AreaCreationEvent;
import io.netty.buffer.ByteBuf;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.world.WorldServer;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.network.ByteBufUtils;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

public class MessageCreateArea implements IMessage
{
    public Area area;

    public MessageCreateArea() {}

    public MessageCreateArea(Area area)
    {
        this.area = area;
    }

    @Override
    public void fromBytes(ByteBuf buf)
    {
        area = new Area(ByteBufUtils.readTag(buf));
    }

    @Override
    public void toBytes(ByteBuf buf)
    {
        ByteBufUtils.writeTag(buf, area.serializeNBT());
    }

    public static class Handler implements IMessageHandler<MessageCreateArea, MessageCreateAreaReply>
    {
        public Handler() {}

        @Override
        public MessageCreateAreaReply onMessage(MessageCreateArea message, MessageContext ctx)
        {
            ((WorldServer) ctx.getServerHandler().player.world).addScheduledTask(new Runnable()
            {
                @Override
                public void run()
                {
                    AddAreaResult result = AddAreaResult.INVALID;
                    Area area = message.area;
                    EntityPlayerMP player = ctx.getServerHandler().player;
                    if(area.getDimensionId() == player.dimension &&
                            area.getMinPos().getY() >= 0 &&
                            area.getMaxPos().getY() <= player.world.getHeight())
                    {
                        CapabilityAreas cap = player.world.getCapability(LandManager.CAPABILITY_AREAS, null);
                        if(cap != null)
                        {
                            //Validation
                            if(!Area.validateName(area.getName()))
                                result = AddAreaResult.INVALID_NAME;
                            else if(cap.hasArea(area.getName()))
                                result = AddAreaResult.NAME_EXISTS;
                            else if(cap.intersectsAnArea(area))
                                result = AddAreaResult.AREA_INTERSECTS;
                            else if(!MinecraftForge.EVENT_BUS.post(new AreaCreationEvent(area)))
                            {
                                //Add new area
                                boolean success = cap.addArea(area);
                                if(success)
                                {
                                    result = AddAreaResult.SUCCESS;
                                    LandManager.areaLog(AreaLogType.CREATE, area.getName(), player);
                                }
                                else
                                    result = AddAreaResult.NAME_EXISTS;
                            }
                        }
                    }

                    LandManager.NETWORK.sendTo(new MessageCreateAreaReply(area.getName(), result), player);
                }
            });
            return null;
        }
    }
}

File: src/main/kotlin/brightspark/landmanager/message/MessageCreateArea.kt
package brightspark.landmanager.message

import brightspark.landmanager.AreaCreationEvent
import brightspark.landmanager.LandManager
import brightspark.landmanager.data.areas.AddAreaResult
import brightspark.landmanager.data.areas.Area
import brightspark.landmanager.util.AreaChangeType
import brightspark.landmanager.util.Message
import brightspark.landmanager.util.areasCap
import brightspark.landmanager.util.sendToPlayer
import net.minecraft.network.PacketBuffer
import net.minecraft.world.server.ServerWorld
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.fml.network.NetworkEvent
import java.util.function.Supplier

class MessageCreateArea : Message {
	private lateinit var area: Area

	@Suppress("unused")
	constructor()

	constructor(area: Area) {
		this.area = area
	}

	override fun encode(buffer: PacketBuffer): Unit = buffer.run {
		writeCompoundTag(area.serializeNBT())
	}

	override fun decode(buffer: PacketBuffer): Unit = buffer.run {
		area = Area(readCompoundTag()!!)
	}

	override fun consume(context: Supplier<NetworkEvent.Context>) {
		context.get().enqueueWork {
			val player = context.get().sender ?: return@enqueueWork
			val world = player.world
			var result = AddAreaResult.INVALID
			if (area.dim == player.world.dimensionKey.location && area.minPos.y >= 0 && area.maxPos.y <= world.height) {
				val cap = world.areasCap
				when {
					!Area.validateName(area.name) -> result = AddAreaResult.INVALID_NAME
					cap.hasArea(area.name) -> result = AddAreaResult.NAME_EXISTS
					cap.intersectsAnArea(area) -> result = AddAreaResult.AREA_INTERSECTS
					!MinecraftForge.EVENT_BUS.post(AreaCreationEvent(area)) -> {
						result = if (cap.addArea(area)) AddAreaResult.SUCCESS else AddAreaResult.NAME_EXISTS
						if (result == AddAreaResult.SUCCESS)
							LandManager.areaChange(
								(world as ServerWorld).server,
								AreaChangeType.CREATE,
								area.name,
								player
							)
					}
				}
			}
			LandManager.NETWORK.sendToPlayer(MessageCreateAreaReply(area.name, result), player)
		}
	}
}
